File: go/day3part2.go
package main
/*
 * day1part2.go - AOC 2021 Day 3 Part 2 in Go.
 * Implements same basic algorithm as Python version, see there for
 * explanations.
 */

import(
  "fmt"
  "os"
  "log"
  "bufio"
  "strconv"
)

// ---------------------------------------------------------------------------
func readFileToBinaryArray(filename string) [][]int {
  var binNums [][]int

  fh,err := os.Open(filename)
  if err != nil {
    log.Fatal(err)
  }
  defer fh.Close()

  s := bufio.NewScanner(fh)
  for s.Scan() {
    t := s.Text()
    var numArray []int
    for _,v := range(t) {
      n,_ := strconv.Atoi(string(v))
      numArray = append(numArray,n)
    }
    binNums = append(binNums, numArray)
  }
  return binNums
}
// -------------------------------------------------------------------------
func getMostPopularValAt(inList [][]int, ndx int) int {
  bitSum := 0
  cutoff := len(inList) / 2
  //Go integer division truncates, not rounds.
  if cutoff * 2 < len(inList) {
    cutoff += 1
  }
  for _,v := range(inList) {
    bitSum += v[ndx]
  }

  if bitSum >= cutoff {
    return 1
  } else {
    return 0
  }
}

// --------------------------------------------------------------------------
func filterArrayByVal(inNumbers [][]int, ndx int, val int) [][]int {
  var outNumbers [][]int
  for _,num := range(inNumbers) {
    if(num[ndx] == val) {
      outNumbers = append(outNumbers,num)
    }
  }
  return outNumbers
}

// -------------------------------------------------------------------------
// do this per-number; technically less efficient (more func calls)
// but the code's easier to write
func calcOxyRating(inNumbers [][]int, pos int) []int {
    selectNum := getMostPopularValAt(inNumbers,pos)

    outNumbers := filterArrayByVal(inNumbers,pos,selectNum)
    if(pos > len(inNumbers[0])) {
      log.Fatal("Failed to determine single filter value after iterating over all positions")
    }
    if(len(outNumbers)==1) {
      return outNumbers[0]
    } else {
      return calcOxyRating(outNumbers,pos+1)
    }
}
// -------------------------------------------------------------------------
func calcCO2Rating(inNumbers [][]int, pos int) []int {
    selectNum := getMostPopularValAt(inNumbers,pos) - 1
    if(selectNum < 0) {
      selectNum = 1
    }
    outNumbers := filterArrayByVal(inNumbers,pos,selectNum)
    if(pos > len(inNumbers[0])) {
      log.Fatal("Failed to determine single filter value after iterating over all positions")
    }
    if(len(outNumbers)==1) {
      return outNumbers[0]
    } else {
      return calcCO2Rating(outNumbers,pos+1)
    }
}

// -------------------------------------------------------------------------
func binArrayToDecimal(binArray []int) int {
  decimal := 0
  for i,v := range(binArray) {
    pow := len(binArray) - i -1;
    decimal += calc2Power(2,pow) * v
  }
  return decimal
}

// -------------------------------------------------------------------------
// there is math.Pow() but that involves float64 conversions....
func calc2Power(base int, exponent int) int {
  result := 1
  for exponent > 0 {
    result *= base
    exponent -= 1
  }
  return result
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------
func main() {
  //var inputfile = "data/day3test.txt"
  var inputfile = "data/day3part1.txt"
  inputNumbers := readFileToBinaryArray(inputfile)

  oxyRating := calcOxyRating(inputNumbers,0)
  co2Rating := calcCO2Rating(inputNumbers,0)
  oxyNum := binArrayToDecimal(oxyRating)
  co2Num := binArrayToDecimal(co2Rating)
  fmt.Printf("Oxygen Generator Rating: %d %v CO2 Scrubber Rating: %d %v\n",oxyNum,oxyRating,co2Num,co2Rating)
  fmt.Printf("LIFE SUPPORT RATING: %d\n", oxyNum * co2Num)
}

File: python/day3part2.py
#!/usr/bin/python3
# Python solution for AOC 2021 Day 3, Part 1
# Given an input of 5-bit binary numbers, calculate "oxygen" and "co2 scrubber"
# values and use that to determine life support rating.
#
# This uses a filtering process - described inline below - to extract the
# required numbers from the input string.

#inputfile="data/day3test.txt"
inputfile="data/day3part1.txt"

##########################################################################
# Find the most popular value (0,1) at position X in an array of binary numbers
def findMostPopularVal(numarray,pos) :
    halfval = len(numarray) / 2
    bitsum=0
    for num in numarray:
        bitsum += num[pos]
    if bitsum >= halfval :
        return 1
    else :
        return 0

###########################################################################
#Load the input stream into an array of binary numbers (as an array)
def buildBinaryArrayFromFile(filename) :
    valarray=[]
    with open(inputfile,"r") as readings:
        for bitval in readings:
            numArray = []
            for i in range(0,len(bitval)) :
                if bitval[i] == '1' :
                    numArray.append(1)
                elif bitval[i] == '0' :
                    numArray.append(0)
            valarray.append(numArray)
    return valarray

###########################################################################
# util func to filter array by val in pos x
def filterArrayByVal(inArray,val,bitPos) :
    outArray=[]
    for num in inArray:
        if num[bitPos] == val :
            outArray.append(num)
    return outArray

###########################################################################
def binNumArrayToDecimal(binArray) :
    binNum=0
    for i in range(len(binArray),0,-1) :
        bit=binArray[i-1]
        pow=len(binArray)-i
        bitval= 2**pow * bit
        binNum+=bitval
        #print(f"{binArray}({i}) bit {bit} dec {bitval} sum: {binNum}")
    return binNum

###########################################################################
# A recursive function to calculate the oxygen generator rating.
# It's a fancy form of filtering with a STOP condition when the inNumbers
# array has only a single member
def findOxyRating(inNumbers, ndx) :
    selectVal = findMostPopularVal(inNumbers,ndx)
    filteredNumbers = filterArrayByVal(inNumbers,selectVal,ndx)

    if len(filteredNumbers) == 1 :
        #FOUND IT!
        return filteredNumbers[0]
    elif ndx == len(filteredNumbers[0]) :
        print(f"ERROR filtering numbers for findOxyRating - reached end but still have {len(filteredNumbers)} candidates: {filteredNumbers}")
    else :
        return findOxyRating(filteredNumbers,ndx+1)

###########################################################################
# I know I could coalesce these into 1 recursive call with appropriate
# filtering but Time Is Money.
def findCO2Rating(inNumbers, ndx) :
    selectVal = abs(findMostPopularVal(inNumbers,ndx)-1)
    filteredNumbers = filterArrayByVal(inNumbers,selectVal,ndx)

    if len(filteredNumbers) == 1 :
        #FOUND IT!
        return filteredNumbers[0]
    elif ndx == len(filteredNumbers[0]) :
        print(f"ERROR filtering numbers for findCO2Rating - reached end but still have {len(filteredNumbers)} candidates: {filteredNumbers}")
    else :
        return findCO2Rating(filteredNumbers,ndx+1)

###########################################################################
###########################################################################
###########################################################################
if __name__ == "__main__" :
    srcNumbers = buildBinaryArrayFromFile(inputfile)

    oxyGenRating = findOxyRating(srcNumbers,0)
    CO2ScrubRating = findCO2Rating(srcNumbers,0)
    o2rat=binNumArrayToDecimal(oxyGenRating)
    co2rat=binNumArrayToDecimal(CO2ScrubRating)
    print(f"Oxygen Generator Rating: {o2rat} {oxyGenRating}, CO2 Scrubber Rating: {co2rat} {CO2ScrubRating}\nLIFE SUPPORT RATING: {o2rat*co2rat}")
